Add Create, Control, Observe introduction guides (#370)
onkernel/docs: introduction/create.mdx

---
title: "Create"
description: "Spin up a cloud browser for your agent"
---

> Kernel browsers are sandboxed Chromium instances that boot in under 30ms. Your agent creates them on demand, drives them, and tears them down — no infra to provision, no warm pool to manage.

Reviewer: ```suggestion
Kernel browsers are sandboxed Chromium instances that boot in under 30ms. Your agent creates them on demand, drives them, and tears them down — no infra to provision, no servers to run.
```

## Your first browser

<CodeGroup>
```typescript Typescript/Javascript
import Kernel from '@onkernel/sdk';

const kernel = new Kernel();

const kernelBrowser = await kernel.browsers.create();
console.log(kernelBrowser.session_id);
```

```python Python
from kernel import Kernel

kernel = Kernel()

kernel_browser = kernel.browsers.create()
print(kernel_browser.session_id)
```

```bash CLI
kernel browsers create

# or with a starting URL and stealth
kernel browsers create --stealth --start-url https://example.com
```
</CodeGroup>

The response includes everything you need to drive the browser: `session_id`, `cdp_ws_url`, `webdriver_ws_url`, and `browser_live_view_url`.

## Pick the right shape

Most of what you'll tune at creation time falls into four buckets:

<Columns cols={2}>
  <Card title="Headless vs headful" href="/browsers/headless">

>     Headful (the default) supports live view and replays. Headless is lighter (1 GB vs 8 GB) and faster — good for short-lived or highly concurrent jobs.

Reviewer: ```suggestion
    headful (default) supports live view, replays, and better stealth — ideal for agent workflows on bot-detected sites. headless is lighter (1 gb vs 8 gb), good for simple scraping.
```

  </Card>
  <Card title="Stealth and proxies" href="/browsers/bot-detection/overview">
    Turn on stealth mode and route through residential, ISP, or datacenter proxies when you're hitting sites with bot detection.
  </Card>
  <Card title="GPU acceleration" href="/browsers/gpu-acceleration">
    Required for WebGL, video, and canvas-heavy workloads. Trades off standby support.
  </Card>
  <Card title="Profiles and auth" href="/auth/overview">
    Persist cookies, storage, and logged-in sessions across runs with a [profile](/auth/profiles), or hand auth off to Kernel entirely with [managed auth](/auth/overview).
  </Card>
</Columns>

## Lifecycle

> A browser stays alive as long as something is driving it — a CDP or WebDriver client, a [Live View](/browsers/live-view) viewer, or an in-flight [computer controls](/browsers/computer-controls) request. After five seconds with none of those active, it enters [standby](/browsers/standby) — state is preserved, billing stops. After the configurable timeout (60s by default) it's deleted.

Reviewer: ```suggestion
A browser stays alive as long as something is driving it — a CDP or WebDriver client, a [Live View](/browsers/live-view) viewer, or an in-flight [computer controls](/browsers/computer-controls) request. After five seconds with none of those active, it enters [standby](/browsers/standby) — state is preserved, billing stops. Once in standby, after the configurable timeout (60s by default) elapses it's deleted.
```

> You can also delete a browser explicitly when you're done:

Reviewer: ```suggestion
We recommend you delete a browser explicitly when you're done with it:
```

<CodeGroup>
```typescript Typescript/Javascript
await kernel.browsers.deleteByID(kernelBrowser.session_id);
```

```python Python
kernel.browsers.delete_by_id(kernel_browser.session_id)
```

```bash CLI
kernel browsers delete <session-id>
```
</CodeGroup>

See [Termination & timeouts](/browsers/termination) for the full set of teardown options.

## Full example

Putting it together — create, connect over CDP, do work, tear down:

<Info>

>   Kernel browsers launch with a default context and page. Make sure to access the existing context and page (`contexts()[0]` and `pages()[0]`) rather than creating a new one.

Reviewer: I feel like one of the goals with this PR is to also promote playwright execution API instead of a direct playwright connection. I think it's worth showing that example here instead of a direct playwright connection. Thoughts?
